### PR TITLE
[monitoring-applications] Refactoring pt. 1

### PR DESCRIPTION
**Pull request by nabokihms**
Signed-off-by: m.nabokikh <maksim.nabokikh@flant.com>

## Description
* Change the module structure so that devs only need to add a folder to add an application.
* Use a single service monitor for all applications to reduce Prometheus CPU usage

## Why do we need it, and what problem does it solve?
This is a preparation to make adding new applications more manageable.

## Checklist
- [x] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: monitoring-applications
type: chore
summary: Change the module structure so that devs only need to add a folder to add an application.
impact_level:  low
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/test'...

**Review by z9r5 — COMMENTED** [review]
If I got this right, the module is still hidden...

**Comment by nabokihms** [comment]
@z9r5 for now, yes, because it is only the first part of refactoring 😅

File: ee/fe/modules/340-monitoring-applications/templates/base-service-monitor.yaml [hunk 1]
```diff
@@ -22,27 +22,27 @@
...
 - sourceLabels: [__meta_kubernetes_endpoint_port_name]
   regex: "https-metrics"
   action: drop
 - sourceLabels:
-  - __meta_kubernetes_service_annotationpresent_prometheus_deckhouse_io_port
-  - __meta_kubernetes_service_annotationpresent_prometheus_deckhouse_io_tls
-  - __meta_kubernetes_endpoint_port_name
+    - __meta_kubernetes_service_annotationpresent_prometheus_deckhouse_io_port
+    - __meta_kubernetes_service_annotationpresent_prometheus_deckhouse_io_tls
+    - __meta_kubernetes_endpoint_port_name
```
Comment (RomanenkoDenys) [inline ee/fe/modules/340-monitoring-applications/templates/base-service-monitor.yaml:33]: Why we move this one step right ? I think it is Goland's work.

File: ee/fe/modules/340-monitoring-applications/templates/base-service-monitor.yaml [hunk 2]
```diff
@@ -22,27 +22,27 @@
   replacement: $1
   targetLabel: endpoint
 
-# HTTP Scheme
+  # HTTP Scheme
```
Comment (RomanenkoDenys) [inline ee/fe/modules/340-monitoring-applications/templates/base-service-monitor.yaml:25]: Why we move this one step right ? I think it is Goland's work.

File: ee/fe/modules/340-monitoring-applications/templates/base-service-monitor.yaml [hunk 3]
```diff
@@ -22,27 +22,27 @@
...
-  - __meta_kubernetes_service_annotationpresent_prometheus_deckhouse_io_tls
-  - __meta_kubernetes_endpoint_port_name
+    - __meta_kubernetes_service_annotationpresent_prometheus_deckhouse_io_port
+    - __meta_kubernetes_service_annotationpresent_prometheus_deckhouse_io_tls
+    - __meta_kubernetes_endpoint_port_name
   regex: "^true;;(.*)|;;http-metrics$"
   action: keep
 
-# HTTPS Scheme
+  # HTTPS Scheme
```
Comment (RomanenkoDenys) [inline ee/fe/modules/340-monitoring-applications/templates/base-service-monitor.yaml:37]: Why we move this one step right ? I think it is Goland's work.

File: ee/fe/modules/340-monitoring-applications/templates/base-service-monitor.yaml [hunk 4]
```diff
@@ -22,27 +22,27 @@
...
 - sourceLabels: [__meta_kubernetes_endpoint_port_name]
   regex: "http-metrics"
   action: drop
 - sourceLabels:
-  - __meta_kubernetes_service_annotationpresent_prometheus_deckhouse_io_port
-  - __meta_kubernetes_service_annotationpresent_prometheus_deckhouse_io_tls
-  - __meta_kubernetes_endpoint_port_name
+    - __meta_kubernetes_service_annotationpresent_prometheus_deckhouse_io_port
+    - __meta_kubernetes_service_annotationpresent_prometheus_deckhouse_io_tls
+    - __meta_kubernetes_endpoint_port_name
```
Comment (RomanenkoDenys) [inline ee/fe/modules/340-monitoring-applications/templates/base-service-monitor.yaml:45]: Why we move this one step right ? I think it is Goland's work.

**Review by RomanenkoDenys — CHANGES_REQUESTED** [review]
LGTM, but i have small suggestions.

File: ee/fe/modules/340-monitoring-applications/hooks/discovery.go [hunk 5]
```diff
@@ -7,36 +7,34 @@ package hooks
...
-	"redis",
-	"sidekiq",
-	"trickster",
-	"grafana",
-	"uwsgi",
-)
+func getAllowedApplications() (set.Set, error) {
+	appsFolder := "../applications/"
+
+	res, err := filepath.Glob(appsFolder + "*")
```
Comment (RomanenkoDenys) [inline ee/fe/modules/340-monitoring-applications/hooks/discovery.go:26]:
```suggestion
	res, err := filepath.Glob("../applications/*")
```

File: ee/fe/modules/340-monitoring-applications/hooks/discovery.go [hunk 6]
```diff
@@ -7,36 +7,34 @@ package hooks
...
+	if err != nil {
+		return nil, err
+	}
+
+	apps := set.New()
+	for _, match := range res {
+		apps.Add(strings.TrimPrefix(match, appsFolder))
```
Comment (RomanenkoDenys) [inline ee/fe/modules/340-monitoring-applications/hooks/discovery.go:33]:
```suggestion
	for _, match := range res {
		apps.Add(filepath.Base(match))
```

File: ee/fe/modules/340-monitoring-applications/templates/base-service-monitor.yaml (hunk 3 above)
Comment (nabokihms) [reply ee/fe/modules/340-monitoring-applications/templates/base-service-monitor.yaml:37]: This one I moved this one intentionally to be in line with the condition.

File: ee/fe/modules/340-monitoring-applications/templates/base-service-monitor.yaml (hunk 2 above)
Comment (nabokihms) [reply ee/fe/modules/340-monitoring-applications/templates/base-service-monitor.yaml:25]: This one I moved this one intentionally to be in line with the condition.

**Review by RomanenkoDenys — APPROVED** [review]
Good work!

**Pull request merged**